src/tracker/client.rs

```diff
@@ -0,0 +1,268 @@
+
+    let req = connect::Request {
+      protocol_id: Self::UDP_TRACKER_MAGIC,
+      action: 0x0000,
+      transaction_id: rng.gen(),
+    };
+
+    let mut buf = [0u8; mem::size_of::<connect::Response>()];
+    let (resp, _) = self.roundtrip(&req, &mut buf)?;
+
+    if resp.transaction_id != req.transaction_id || resp.action != req.action {
+      return Err(Error::TrackerResponse);
+    }
+
+    self.state = State::Connected {
+      id: resp.connection_id,
+      epoch: Instant::now(),
+      socket_addr: self.sock.peer_addr().context(error::Network)?,
+      host_port: host_port.clone(),
+    };
+
+    Ok(())
+  }
+
+  pub fn announce(&self, btinh: Infohash) -> Result<Vec<SocketAddr>> {
+    let mut rng = rand::thread_rng();
+    let req = announce::Request {
+      connection_id: self.state.get_connection_id()?,
+      action: 0x0001,
+      transaction_id: rng.gen(),
+      infohash: btinh.into(),
+      peer_id: self.peer_id,
+      downloaded: 0x0000,
+      left: u64::MAX,
+      uploaded: 0x0000,
+      event: 0x0000,
+      ip_address: 0x0000,
+      num_want: u32::MAX,
+      port: self.sock.local_addr().context(error::Network)?.port(),
+    };
+    let mut buf = [0u8; 8192]; // is there a BUFSIZ macro for Rust?
+    let (resp, len) = self.roundtrip(&req, &mut buf)?;
+
+    if resp.transaction_id != req.transaction_id || resp.action != req.action {
+      return Err(Error::TrackerResponse);
+    }
+    self.parse_compact_peer_list(&buf[mem::size_of::<announce::Response>()..len])
+  }
+
+  fn roundtrip<T: Request>(&self, req: &T, rxbuf: &mut [u8]) -> Result<(T::Response, usize)> {
+    let msg = req.serialize();
+    let read = self.send_and_retry_with_backoff(&msg, rxbuf)?;
+
+    if read == 0 {
+      Err(Error::TrackerTimeout)
+    } else {
+      T::Response::deserialize(&rxbuf[..read])
+    }
+  }
+
+  fn send_and_retry_with_backoff(&self, txbuf: &'a [u8], rxbuf: &'a mut [u8]) -> Result<usize> {
+    let mut len_read: usize = 0;
+
+    for attempt in 0..=8 {
+      self.sock.send(txbuf).context(error::Network)?;
+      self
+        .sock
+        .set_read_timeout(Some(Duration::new(15 * 2u64.pow(attempt), 0)))
+        .context(error::Network)?;
+
+      if let Ok(len) = self.sock.recv(rxbuf) {
+        len_read = len;
+        break;
+      } else {
+        // Abort sending if the connection id goes stale during a back off
+        if let State::Connected { epoch, .. } = self.state {
+          if epoch.elapsed().as_secs() > 60 {
+            return Err(Error::TrackerConnectionExpired);
+          }
+        }
+      }
+    }
+
+    if len_read == 0 {
+      Err(Error::TrackerTimeout)
+    } else {
+      Ok(len_read)
+    }
+  }
+
+  // XXX: perhaps this should be in a different namespace
+  fn parse_compact_peer_list(&self, buf: &[u8]) -> Result<Vec<SocketAddr>> {
+    let mut subswarm = Vec::<SocketAddr>::new();
+
+    let stride = if self.state.is_ipv6() { 18 } else { 6 };
+
+    // XXX: consider array_chunks()
```

Comment:
Since array_chunks is nightly, and probably will be for a while, you can omit the note. And, actually, it wouldn't work here, since `stride` isn't a const.